tests/common/graceful_shutdown_test.go

```diff
@@ -0,0 +1,87 @@
+// Copyright 2023 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build e2e
+
+package common
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+
+	"go.etcd.io/etcd/tests/v3/framework/config"
+	"go.etcd.io/etcd/tests/v3/framework/interfaces"
+)
+
+func TestGracefulShutdown(t *testing.T) {
+	testRunner.BeforeTest(t)
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	clus := testRunner.NewCluster(ctx, t, config.WithClusterSize(3))
+	// clean up orphaned resources like closing member client.
+	defer clus.Close()
+	// shutdown each etcd member process sequentially
+	// and start from old leader, (new leader), (follower)
+	members := clus.Members()
+	mn := len(members)
+	for len(members) > 0 {
+		// transfer leadership only when quorum is active after leader shuts down
+		expectTransferLeadership := len(members)-1 >= 1+(mn-1)/2
+		members = tryShutdownLeader(ctx, t, members, expectTransferLeadership)
+	}
+}
+
+// tryShutdownLeader tries stop etcd member if it is leader. If no leader is found, stop the very first member.
+// it also asserts stop member should not take longer than 1.5 seconds and lead ID has been changed.
+func tryShutdownLeader(ctx context.Context, t *testing.T, members []interfaces.Member, expectTransferLeadership bool) (remainingMembers []interfaces.Member) {
+	var leaderID uint64
+	leaderIdx := -1
+	for i, m := range members {
+		mc := m.Client()
+		sresps, err := mc.Status(ctx)
+		require.NoError(t, err)
+		if sresps[0].Leader == sresps[0].Header.MemberId {
+			leaderIdx = i
+			leaderID = sresps[0].Leader
+			stopMember(t, m)
+			break
+		}
+	}
+
+	if leaderIdx != -1 {
+		remainingMembers = append(members[:leaderIdx], members[leaderIdx+1:]...)
+		// validate leader ID has been changed after stopMember
+		resps, err := remainingMembers[0].Client().Status(ctx)
+		require.NoError(t, err)
+		if expectTransferLeadership {
+			require.NotEqualf(t, resps[0].Leader, leaderID, "expect old leaderID %x changed to new leader ID %x", leaderID, resps[0].Leader)
```

Comment:
Note, leadership transfer will happen even no matter if leader does it gracefully. Question is how much time it takes. I would add `require.Less` block similar to `stopMember`

Reply:
Good point! Thanks! 

I would like adding this in the test assertion in the next revision.